tests/transport/rtps_direct_write/publisher.cpp

```diff
@@ -0,0 +1,305 @@
+    size_t size = 0, padding = 0;
+    gen_find_size(locators, size, padding);
+    ACE_Message_Block mb(size + padding + 1);
+    Serializer ser(&mb, ACE_CDR_BYTE_ORDER, Serializer::ALIGN_CDR);
+    ser << locators;
+    ser << ACE_OutputCDR::from_boolean(false);
+    OpenDDS::RTPS::message_block_to_sequence(mb, data);
+  }
+
+  const Domain& domain;
+  ssize_t callbacks_expected_;
+};
+
+class DDS_TEST {
+public:
+  DDS_TEST(int argc, ACE_TCHAR* argv[]) : domain(argc, argv), msgSeqN(0) {
+    domain.setHeartbeatPeriod(100);
+
+    ACE_INET_Addr remoteAddr = domain.getHostAddress();
+    locators.length(1);
+    locators[0].kind = OpenDDS::RTPS::address_to_kind(remoteAddr);
+    locators[0].port = remoteAddr.get_port_number();
+    OpenDDS::RTPS::address_to_bytes(locators[0].address, remoteAddr);
+  }
+
+  int run();
+
+private:
+  static const bool hostIsBigEndian = !ACE_CDR_BYTE_ORDER;
+  static const ACE_CDR::ULong encap = 0x00000100; // {CDR_LE, options} in BE format
+  static const CORBA::Octet DE  = OpenDDS::RTPS::FLAG_D | OpenDDS::RTPS::FLAG_E;
+  static const CORBA::Octet DEQ = OpenDDS::RTPS::FLAG_D | OpenDDS::RTPS::FLAG_E | OpenDDS::RTPS::FLAG_Q;
+
+  void force_inline_qos(bool val) {
+    OpenDDS::DCPS::RtpsUdpDataLink::force_inline_qos_ = val;
+  }
+  void log_time(const ACE_Time_Value& t) const {
+    ACE_TCHAR buffer[32];
+    const std::time_t seconds = t.sec();
+    std::string ts(ACE_TEXT_ALWAYS_CHAR(ACE_OS::ctime_r(&seconds, buffer, 32)));
+    ts.erase(ts.size() - 1); // remove \n from ctime()
+    ACE_DEBUG((LM_INFO, "Sending with timestamp %C %q usec\n", ts.c_str(), ACE_INT64(t.usec())));
+  }
+
+  bool writeToSocket(const TestMsg& msg, const CORBA::Octet flags = DE) const;
+  void serializeData(DataSampleElement& ds, const TestMsg& msg) const;
+
+  void listData(SendStateDataSampleList& list, DataSampleElement ds[], const ssize_t size) {
+    list.head_ = ds;
+    list.size_ = size;
+    list.tail_ = &ds[size - 1];
+    for (int i = 0; i < size - 1; ++i) {
+      ds[i].set_next_send_sample(&ds[i + 1]);
+    }
+  }
+
+  Domain domain;
+  LocatorSeq locators;
+  mutable CORBA::ULong msgSeqN;
+};
+
+using namespace OpenDDS::RTPS;
+
+bool DDS_TEST::writeToSocket(const TestMsg& msg, const CORBA::Octet flags) const
+{
+  const OpenDDS::RTPS::GuidPrefix_t& local_prefix = domain.getPubWtrId().guidPrefix;
+  const Header hdr = { {'R', 'T', 'P', 'S'}, PROTOCOLVERSION, VENDORID_OPENDDS,
+    {local_prefix[0], local_prefix[1], local_prefix[2], local_prefix[3],
+     local_prefix[4], local_prefix[5], local_prefix[6], local_prefix[7],
+     local_prefix[8], local_prefix[9], local_prefix[10], local_prefix[11]} };
+
+  const ACE_Time_Value now = ACE_OS::gettimeofday();
+  log_time(now);
+  const double conv = 4294.967296; // NTP fractional (2^-32) sec per microsec
+  const InfoTimestampSubmessage it = { {INFO_TS, 1, 8},
+    {static_cast<ACE_CDR::Long>(now.sec()),
```

Comment:
This line was an error on Azure CI, probably due to the Time_t changes from another branch.  This may be a case where explicitly merging master into this branch is the best way forward.

Reply:
There are a few related errors on the Azure CI Linux/clang builds (even with the latest revision 4829273)